Generate only one Type class per backend

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19295 Generate only one Type class per backend**
* #19275 Make complex its own backend
* #19270 Add ScalarType argument to Type::options()
* #19230 Generate cases for all ScalarTypes in Type functions that call to TH
* #19183 Remove usages of TypeID
* #18972 Make copy_ a native function, remove copy and copy_ from Type

- Generate a single Type for each backend, we've made them all generate the same functions in #19230 
- Change Type registration to take a single Type for each backend
- Type lookup functions like getNonVariableType() still take ScalarType for now, but it's unused.

Differential Revision: [D14948581](https://our.internmc.facebook.com/intern/diff/D14948581)

Commit:
Generate only one Type class per backend

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

smessmer reviewed (approved):
yay :)

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

dzhulgakov commented on aten/src/ATen/core/Type.h:
```diff
@@ -45,64 +45,14 @@ struct Quantizer;
+  QuantizedCPU,
+  MSNPU,
+  XLA,
   CPUComplexFloat,
```
what's the story with Complex btw? - why there are still switches based on scalar type?

dzhulgakov commented on aten/src/ATen/core/Type.h:
```diff
@@ -45,64 +45,14 @@ struct Quantizer;
 static inline void noop_deleter(void*) {}
 
 enum class TypeID {
```
does it match TensorTypeID now precisely?

li-roy replied:
No, there are some TensorTypeIDs that are unique to caffe2.

dzhulgakov commented on aten/src/ATen/gen.py:
```diff
@@ -451,19 +442,19 @@ def declare_outputs():
         if density == 'Dense':
-            fm.will_write("{}{}{}{}.h".format('LegacyTH', full_backend, scalar_name, 'Dispatcher'))
-            fm.will_write("{}{}{}{}.cpp".format('LegacyTH', full_backend, scalar_name, 'Dispatcher'))
+            fm.will_write("{}{}{}{}.h".format('LegacyTH', backend, scalar_type[0], 'Dispatcher'))
```
what is this LegacyTHDispatcher? it seems to be a bunch of empty cpp classes. Can't we just nuke all of that now?

gchanan replied:
@li-roy did you answer this?

This was meant to move all the TH/THNN functions to a different mechanism, primarily because they aren't end-user methods (and don't have a schema).  The unsolved problem is how to handle autogrrad: they still need variable type definitions that know calls to them need to be unwrapped.

dzhulgakov reviewed (commented):
this is awesome!

ezyang commented on test/cpp_extensions/complex_registration_extension.cpp:
```diff
@@ -66,8 +66,7 @@ struct CPUComplexFloatType : public at::CPUTypeDefault {
   void registerComplexTypes(Context* context) const override {
-    context->registerType(
-        Backend::ComplexCPU, ScalarType::ComplexFloat, new CPUComplexFloatType());
+    context->registerType(Backend::ComplexCPU, new CPUComplexFloatType());
```
If you're killing per-dtype Types, you should rename CPUComplexFloatType to CPUComplexType

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

Commit:
Update on "Generate only one Type class per backend"

Generate only one Type class per backend

gh-metadata: pytorch pytorch 19295 gh/li-roy/11/head

dzhulgakov commented on aten/src/ATen/core/Type.h:
```diff
@@ -45,64 +45,14 @@ struct Quantizer;
+  MSNPU,
+  XLA,
   CPUComplexFloat,
   CPUComplexDouble,
```
@li-roy - why do we still have CPUComplexDouble here if you changed to CPUComplex elsewhere?

li-roy replied:
Oops, I forgot to delete these.

gchanan commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -55,7 +55,7 @@ class CAFFE2_API LegacyTypeDispatch {
   // have a Tensor and thus the Type of that Tensor must already
   // be initialized.
   Type* getNonVariableTypeRaw(Backend p, ScalarType s) {
```
shouldn't this no longer take a ScalarType?

gchanan commented on test/expect/TestScript.test_print-stdout.expect:
```diff
@@ -2,4 +2,4 @@
  0.9975
  0.9999
-[ Variable[CPUDoubleType]{4} ] 1 2 [1, 2] [1., 2.]
+[ Variable[CPUType]{4} ] 1 2 [1, 2] [1., 2.]
```
this seems like a regression in usability.

gchanan reviewed (commented):
This makes some major internal changes, but has no written justification in the commit message.  Is it written somewhere?
